neverpile-commons-authorization/src/main/java/com/neverpile/common/authorization/policy/impl/PolicyBasedAuthorizationService.java

```diff
@@ -66,23 +72,79 @@ public boolean isAccessAllowed(final String resourceSpecifier, final Set<Action>
+        .filter(r -> satisfiesConditions(r, context)) //
+        .collect(Collectors.toList());
+
+    // match each action individually
+    boolean allMatched = true;
+    for (Action a : actions) {
+      Optional<AccessRule> matchingRule = matchingRules.stream() //
+          .filter(r -> matchesActions(a.key(), r.getActions())) //
+          .findFirst();
+      if (matchingRule.isPresent() && matchingRule.get().getEffect() == Effect.DENY) {
+        // deny means deny
+        return false;
+      }
+      allMatched &= matchingRule.isPresent();
+    }
+
+    // evaluate applicable effect: if we had explicit matches for all rules, we allow, otherwise we
+    // revert to the default effect
+    Effect e = allMatched ? Effect.ALLOW : //
+        policy.getDefaultEffect() != null ? policy.getDefaultEffect() : Effect.DENY;
 
     LOGGER.debug("Authorization for {} on {} with principal {}: {}", actions, resourceSpecifier,
         authentication.isAuthenticated() ? authentication.getPrincipal() : "anonymous", e);
 
     return e == Effect.ALLOW;
   }
 
+  @Override
+  public List<Permission> getPermissions(final String resourceSpecifier, final AuthorizationContext context) {
+    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+    // Collection the permissions into a list optimizing the number of permissions
+    // by combining successive permissions with the same effect into one.
+    return getPermissions(resourceSpecifier, context, authentication) //
+        .collect(ArrayList::new, (l, p) -> {
+          if (!l.isEmpty()) {
+            Permission last = l.get(l.size() - 1);
+            if (last.getEffect() == p.getEffect()) {
+              ArrayList<String> combined = new ArrayList<>(last.getActionKeys());
+              combined.addAll(p.getActionKeys());
+              last.setActionKeys(combined);
+              return;
+            }
+          }
+          l.add(p);
+        }, ArrayList::addAll);
+  }
+
+  private Stream<Permission> getPermissions(final String resourceSpecifier, final AuthorizationContext context,
+      final Authentication authentication) {
+    AccessPolicy policy = policyRepository.getCurrentPolicy();
+
+    // stream of matching rules converted to permissions
+    Stream<Permission> rulePermissions = policy.getRules().stream() //
+        .filter(null != authentication && authentication.isAuthenticated() //
+            ? (r) -> matchesAuthentication(r, authentication) //
+            : this::matchesAnonymousUser) //
+        .filter(r -> matchesResource(r, resourceSpecifier)) //
+        .filter(r -> satisfiesConditions(r, context)) //
+        .map(r -> new Permission(r.getEffect(), r.getActions()));
+
+    // if the default effect is ALLOW, add a final permission
+    if (policy.getDefaultEffect() == Effect.ALLOW)
+      return Stream.concat(rulePermissions, Stream.of(new Permission(Effect.ALLOW, Arrays.asList(Action.ANY.key()))));

```

Comment:
Man munkelt, die Abkürzung hat die NASA in der Vergangenheit schon Raketen gekostet... 😄 Würde in der Businesslogik auf die Abkürzung verzichten (im Logging 163ff hab ich's auch gesehen - da ist die Produktion aber eher nicht davon betroffen...)
```suggestion
    if (policy.getDefaultEffect() == Effect.ALLOW) {
      return Stream.concat(rulePermissions, Stream.of(new Permission(Effect.ALLOW, Arrays.asList(Action.ANY.key()))));
   }
```